compose/config/config.py

```diff
@@ -513,41 +512,41 @@ def interpolate_config_section(config_file, config, section, environment):
     )
 
 
-def process_config_file(config_file, environment, service_name=None):
-    services = interpolate_config_section(
-        config_file,
-        config_file.get_service_dicts(),
-        'service',
-        environment)
+def process_config_file(config_file, environment, service_name=None, interpolate=True):
```

Comment:
I would suggest this instead, which introduces less changes overall and is easier to isolate in a review:
```diff
diff --git a/compose/config/config.py b/compose/config/config.py
index 24c9c1c5..20e83577 100644
--- a/compose/config/config.py
+++ b/compose/config/config.py
@@ -503,50 +503,63 @@ def load_services(config_details, config_file, compatibility=False):
     return build_services(service_config)
 
 
-def interpolate_config_section(config_file, config, section, environment):
-    return interpolate_environment_variables(
-        config_file.version,
-        config,
-        section,
-        environment
-    )
+def process_config_section(config_file, config, section, environment, interpolate):
+    validate_config_section(config_file.filename, config, section)
+    if interpolate:
+        return interpolate_environment_variables(
+            config_file.version,
+            config,
+            section,
+            environment
+        )
+    else:
+        return config
 
 
 def process_config_file(config_file, environment, service_name=None, interpolate=True):
 
-    def validate(key, val):
-        validate_config_section(config_file.filename, val, key)
-
-    processed_config = {}
-    if interpolate:
-        def process(key, val):
-            processed_config[key+'s'] = interpolate_config_section(
-                config_file,
-                val,
-                key,
-                environment
-            )
-    else:
-        def process(key, val):
-            processed_config[key+'s'] = val
-
-    data = [('service', config_file.get_service_dicts())]
+    services = process_config_section(
+        config_file,
+        config_file.get_service_dicts(),
+        'service',
+        environment,
+        interpolate,
+    )
     if config_file.version > V1:
-        data += ([('volume', config_file.get_volumes()),
-                  ('network', config_file.get_networks())])
+        processed_config = dict(config_file.config)
+        processed_config['services'] = services
+        processed_config['volumes'] = process_config_section(
+            config_file,
+            config_file.get_volumes(),
+            'volume',
+            environment,
+            interpolate,
+        )
+        processed_config['networks'] = process_config_section(
+            config_file,
+            config_file.get_networks(),
+            'network',
+            environment,
+            interpolate,
+        )
         if config_file.version >= const.COMPOSEFILE_V3_1:
-            data += [('secret', config_file.get_secrets())]
+            processed_config['secrets'] = process_config_section(
+                config_file,
+                config_file.get_secrets(),
+                'secret',
+                environment,
+                interpolate,
+            )
         if config_file.version >= const.COMPOSEFILE_V3_3:
-            data += [('config', config_file.get_configs())]
-        processed_config = dict(config_file.config)
-
-    for key, val in data:
-        validate(key, val)
-        process(key, val)
-
-    services = processed_config['services']
-    if config_file.version <= V1:
-        processed_config = processed_config['services']
+            processed_config['configs'] = process_config_section(
+                config_file,
+                config_file.get_configs(),
+                'secret',
+                environment,
+                interpolate,
+            )
+    else:
+        processed_config = services
 
     config_file = config_file._replace(config=processed_config)
     validate_against_config_schema(config_file)

```